README.md
# Itemset Mining
<a href="https://pypi.org/project/itemset-mining/"><img src="https://img.shields.io/pypi/dm/itemset-mining?style=for-the-badge" alt="downloads" /></a>  

<a href="https://pypi.org/project/itemset-mining/"><img src="https://img.shields.io/pypi/v/itemset-mining?style=for-the-badge" alt="latest release" /></a>
<a href="https://pypi.org/project/itemset-mining/"><img src="https://img.shields.io/pypi/pyversions/itemset-mining?style=for-the-badge" alt="supported python versions" /></a>
<a href="https://pypi.org/project/itemset-mining/"><img src="https://img.shields.io/pypi/status/itemset-mining?style=for-the-badge" alt="package status" /></a>
<a href="https://github.com/ZaxR/itemset-mining/blob/master/LICENSE"><img src="https://img.shields.io/pypi/l/itemset-mining?style=for-the-badge" alt="license" /></a>

<a href="https://travis-ci.com/ZaxR/itemset_mining"><img src="https://img.shields.io/travis/com/ZaxR/itemset_mining?style=for-the-badge" alt="travis build status" /></a>
<a href="https://itemset_mining.readthedocs.io/en/latest/"><img src="https://img.shields.io/readthedocs/itemset_mining/latest?style=for-the-badge" alt="docs build status" /></a>
<a href="https://codecov.io/gh/ZaxR/itemset_mining"><img src="https://img.shields.io/codecov/c/github/zaxr/itemset_mining?style=for-the-badge" alt="coverage status" /></a>

Implements itemset mining algorithms.

## Algorithms

### High-utility itemset mining (HUIM)
HUIM generalizes the problem of frequent itemset mining (FIM) by considering item values and weights. A popular application of HUIM is to discover all sets of items purchased together by customers that yield a high profit for the retailer. In such a case, item values would show not just that a load of bread is in a basket, but how many there are; and weights would include the profit from a loaf of bread.

More technically, HUIM requires transactions in the transactions "database" to have internal utilities (i.e. item values) associated with each item in each transaction and a "database" of external utilities for each item (i.e. weights).

| Algorithm        | Class                                  | How to Cite |
|------------------|----------------------------------------|-------------|
| [Two-Phase\*][1] | itemset_mining.two_phase_huim.TwoPhase | [Link][2]   |

\* Includes max length support<br>

[1]: <http://cucis.ece.northwestern.edu/publications/pdf/LiuLia05A.pdf> "Two-Phase (2005)"
[2]: <https://link.springer.com/chapter/10.1007/11430919_79#citeas> "Get citation text"

## Roadmap (high to low priority):
- **Address low-correlation HUIs with one of bond, all-confidence, or affinity.** Itemsets that are high utility, but where the items aren't correlated can be misleading for making marketing decisions. E.g. if an itemset of a TV and a pen is a HUI, it's likely just because the TV is expensive, not because it's an interesting pattern.
- **Add *average* utility measure support**, for easier, more intuitive minutil
- **Support discount strategies** via a discount strategy table and upgraded external utilities table.
- **Add top-k HUI support.**
- **Support identifying periodic high utility itemsets.** This allows detection of purchase patterns among high-utility itemsets to allow cross-promotions to customers who buy sets of items periodically.
- **Support items' on-shelf time.** Ignmoring on-shelf time will biat toward items that have more shelf time, since they have more chance to generate higher utility.
- **Allow incremental transaction updates** without rerunning everything.
- **Support concise HUI itemsets, specifically closed form.** This allows the algorithm to be more efficient, only showing longer itemsets, which may be the most interesting ones (correlation issues aside).

### Installation:
```bash
pip install itemset-mining
```

### Example:

```python
    >>> from operator import attrgetter
    >>> from itemset_mining.two_phase_huim import TwoPhase
    >>> transactions = [
    ...     [("Coke 12oz", 6), ("Chips", 2), ("Dip", 1)],
    ...     [("Coke 12oz", 1)],
    ...     [("Coke 12oz", 2), ("Chips", 1), ("Filet Mignon 1lb", 1)],
    ...     [("Chips", 1)],
    ...     [("Chips", 2)],
    ...     [("Coke 12oz", 6), ("Chips", 1)]
    ... ]
    >>> # ARP for each item
    >>> external_utilities = {
    ...     "Coke 12oz": 1.29,
    ...     "Chips": 2.99,
    ...     "Dip": 3.49,
    ...     "Filet Mignon 1lb": 22.99
    ... }
    >>> # Minimum dollar value generated by an itemset we care about across all transactions
    >>> minutil = 20.00

    >>> hui = TwoPhase(transactions, external_utilities, minutil)
    >>> result = hui.get_hui()
    >>> sorted(result, key=attrgetter('itemset_utility'), reverse=True)
    ... # doctest: +NORMALIZE_WHITESPACE
    [HUIRecord(items=('Chips', 'Coke 12oz'), itemset_utility=30.02),
     HUIRecord(items=('Chips', 'Coke 12oz', 'Filet Mignon 1lb'), itemset_utility=28.56),
     HUIRecord(items=('Chips', 'Filet Mignon 1lb'), itemset_utility=25.979999999999997),
     HUIRecord(items=('Coke 12oz', 'Filet Mignon 1lb'), itemset_utility=25.57),
     HUIRecord(items=('Filet Mignon 1lb',), itemset_utility=22.99),
     HUIRecord(items=('Chips',), itemset_utility=20.93)]

```


CHANGELOG.md
Changelog
==========

<h2>[0.2.2] - 2022-04-23</h2>
**Changed**
- Deduplicate `HUIRecord`s in results when itemset length >= 3

<h2>[0.2.1] - 2020-10-13</h2>
**Changed**
- Hotfix missing VERSION file preventing library import

<h2>[0.2.0] - 2020-08-02</h2>
**Added**
- Support for generator transaction inputs to Two-Phase HUIM algorithm
**Changed**
- Fix examples to pass doctest

<h2>[0.1.0] - 2020-07-23</h2>
**Added**
- Added Two-Phase HUIM algorithm


tests/test_two_phase.py
from operator import attrgetter

import pytest
from pytest_cases import fixture_ref, parametrize

from itemset_mining.two_phase_huim import HUIRecord, TwoPhase


@pytest.fixture(scope="module")
def transactions_list():
    return [
        [("Coke 12oz", 6), ("Chips", 2), ("Dip", 1)],
        [("Coke 12oz", 1)],
        [("Coke 12oz", 2), ("Chips", 1), ("Filet Mignon 1lb", 1)],
        [("Chips", 1)],
        [("Chips", 2)],
        [("Coke 12oz", 6), ("Chips", 1)],
    ]


@pytest.fixture
def transactions_generator(transactions_list):
    return (t for t in transactions_list)


@pytest.fixture(scope="module")
def external_utilities():
    return {
        "Coke 12oz": 1.29,
        "Chips": 2.99,
        "Dip": 3.49,
        "Filet Mignon 1lb": 22.99,
    }


@parametrize(
    "transactions, ext_utilities",
    [
        (fixture_ref(transactions_list), fixture_ref(external_utilities)),
        (fixture_ref(transactions_generator), fixture_ref(external_utilities)),
    ],
)
def test_two_phase(transactions, ext_utilities):

    minutil = 20.00

    hui = TwoPhase(transactions, ext_utilities, minutil)
    unsorted_result = hui.get_hui()
    result = sorted(unsorted_result, key=attrgetter("itemset_utility"), reverse=True)
    expected = [
        HUIRecord(items=("Chips", "Coke 12oz"), itemset_utility=30.02),
        HUIRecord(
            items=("Chips", "Coke 12oz", "Filet Mignon 1lb"), itemset_utility=28.56
        ),
        HUIRecord(
            items=("Chips", "Filet Mignon 1lb"), itemset_utility=25.979999999999997
        ),
        HUIRecord(items=("Coke 12oz", "Filet Mignon 1lb"), itemset_utility=25.57),
        HUIRecord(items=("Filet Mignon 1lb",), itemset_utility=22.99),
        HUIRecord(items=("Chips",), itemset_utility=20.93),
    ]

    assert result == expected


itemset_mining/two_phase_huim.py
from collections import defaultdict, namedtuple
from itertools import chain, combinations, tee
from types import GeneratorType
from typing import Any, Dict, Generator, List, Optional, Set, Tuple, Union

TupleOfStrAndNum = Tuple[str, Union[int, float]]

HUIRecord = namedtuple(  # pylint: disable=C0103
    "HUIRecord", ("items", "itemset_utility")
)


CandidateHUIRecord = namedtuple("HUIRecord", ("items", "twu"))  # pylint: disable=C0103


class TwoPhase:
    """

    Example:
        >>> from itemset_mining.two_phase_huim import TwoPhase
        >>> from operator import attrgetter
        >>> transactions = [
        ...     [("Coke 12oz", 6), ("Chips", 2), ("Dip", 1)],
        ...     [("Coke 12oz", 1)],
        ...     [("Coke 12oz", 2), ("Chips", 1)],
        ...     [("Chips", 1)],
        ...     [("Chips", 2)],
        ...     [("Coke 12oz", 6), ("Chips", 1)]
        ... ]
        >>> # ARP for each item
        >>> external_utilities = {
        ...     "Coke 12oz": 1.29,
        ...     "Chips": 2.99,
        ...     "Dip": 3.49
        ... }
        >>> # Minimum dollar value generated by an itemset we care about across all transactions
        >>> minutil = 20.00

        >>> hui = TwoPhase(transactions, external_utilities, minutil)
        >>> result = hui.get_hui()
        >>> sorted(result, key=attrgetter('itemset_utility'), reverse=True)
        ... # doctest: +NORMALIZE_WHITESPACE
        [HUIRecord(items=('Chips', 'Coke 12oz'), itemset_utility=30.02),
         HUIRecord(items=('Chips',), itemset_utility=20.93)]

    """

    def __init__(
        self,
        transactions: Union[
            Generator[TupleOfStrAndNum, None, None], List[TupleOfStrAndNum]
        ],
        external_utilities: Dict,
        minutil: int,
    ):
        # todo: sort transaction items lexographically?
        self.transactions_type = type(transactions).__name__  # "generator" or "list"
        self.transactions = self.ensure_transaction_items_unique(transactions)

        # For items in transactions, but not in external_utilities,
        # self.external_utilities[item] = 0
        # can still loop through those items for initial candidate set,
        # because if the external utilities are 0, not going to be important anyway
        self.external_utilities = defaultdict(int, external_utilities)
        self.minutil = minutil

        # self.num_transaction = 0
        self._items = list(external_utilities.keys())
        # todo: see if keeping track of item indices offers speedup
        # self._item_transaction_ids = {}

    @staticmethod
    def _ensure_transaction_items_unique_generator(transactions):
        for idx, transaction in enumerate(transactions):
            t_list = list(transaction)
            item_names = [item[0] for item in t_list]
            if len(item_names) != len(set(item_names)):
                raise ValueError(
                    f"There are duplicate items in transaction index {idx}. "
                    f"Transactions must have a unique list of items, "
                    f"with their internal utilities aggregated togther."
                )
            yield transaction

    @staticmethod
    def _ensure_transaction_items_unique_memory(transactions):
        for idx, transaction in enumerate(transactions):
            item_names = [item[0] for item in transaction]
            if len(item_names) != len(set(item_names)):
                raise ValueError(
                    f"There are duplicate items in transaction index {idx}. "
                    f"Transactions must have a unique list of items, "
                    f"with their internal utilities aggregated togther."
                )
        return transactions

    @classmethod
    def ensure_transaction_items_unique(cls, transactions):
        """Ensure there are no duplicate items within a transaction,

        while preserving generators.
        """
        if isinstance(transactions, GeneratorType):
            transactions = cls._ensure_transaction_items_unique_generator(transactions)
        else:
            transactions = cls._ensure_transaction_items_unique_memory(transactions)
        return transactions

    @property
    def items(self):
        """Returns the list of items that appear in transactions."""
        return sorted(self._items)

    def calc_transaction_utility(self, transaction: Tuple[Any, Union[int, float]]):
        return sum(
            [
                self.external_utilities[item] * internal_utility
                for item, internal_utility in transaction
            ]
        )

    def calc_twu(self, itemset: List[str]):
        """Calculated the transaction-weighted utilization for an itemset"""
        twu = 0

        if self.transactions_type == "generator":
            self.transactions, transactions = tee(self.transactions)
        else:
            transactions = self.transactions

        for transaction in transactions:
            transaction_itemset = [t[0] for t in transaction]
            if all(item in transaction_itemset for item in itemset):
                twu += self.calc_transaction_utility(transaction)

        return twu

    def calc_itemset_utility(self, itemset: Tuple):
        itemset_utility = 0

        if self.transactions_type == "generator":
            self.transactions, transactions = tee(self.transactions)
        else:
            transactions = self.transactions

        for transaction in transactions:
            transaction_itemset = [i[0] for i in transaction]
            if all(item in transaction_itemset for item in itemset):
                transaction_part = [
                    tup for item in itemset for tup in transaction if tup[0] == item
                ]
                itemset_utility += self.calc_transaction_utility(transaction_part)

        return itemset_utility

    def initial_candidates(self):
        """Returns the initial candidates."""
        return [tuple([item]) for item in self.items]

    @staticmethod
    def _create_next_candidates(prev_candidates: Set, length: int):
        """Returns the apriori candidates as a list.

        Args:
            prev_candidates: Previous candidates as a list.
            length: The lengths of the next candidates.

        """
        items = sorted(set(chain.from_iterable(prev_candidates)))

        # Create the temporary candidates. These will be filtered below.
        tmp_next_candidates = (itemset for itemset in combinations(items, length))

        # Return all the candidates if the length of the next candidates is 2
        # because their subsets are the same as items.
        if length < 3:
            return tmp_next_candidates

        # Filter to candidates where all of their subsets are in the previous candidates.
        next_candidates = [
            candidate
            for candidate in tmp_next_candidates
            if all(
                itemset in prev_candidates
                for itemset in combinations(candidate, length - 1)
            )
        ]
        return next_candidates

    def get_high_twu_itemsets(
        self, max_length: Optional[int] = None
    ) -> Generator[CandidateHUIRecord, None, None]:
        """Returns a generator of support records with given transactions.

        This is "Phase 1."

        """
        candidate_itemsets = self.initial_candidates()
        length = 1
        while candidate_itemsets:
            high_util_itemsets = set()
            for candidate_itemset in candidate_itemsets:
                twu = self.calc_twu(candidate_itemset)
                if twu < self.minutil:
                    continue
                high_util_itemsets.add(candidate_itemset)
                yield CandidateHUIRecord(candidate_itemset, twu)
            length += 1
            if max_length and length > max_length:
                break
            candidate_itemsets = self._create_next_candidates(
                high_util_itemsets, length
            )

    def get_hui(
        self, max_length: Optional[int] = None
    ) -> Generator[HUIRecord, None, None]:
        # Phase I
        high_twu_itemsets = self.get_high_twu_itemsets(max_length=max_length)
        candidate_itemsets = (itemset.items for itemset in high_twu_itemsets)

        # Phase II
        for itemset in candidate_itemsets:
            itemset_utility = self.calc_itemset_utility(itemset)
            if itemset_utility < self.minutil:
                continue
            """
            Note:
                Making a [frozen]set of items isn't necessary,
                since input is required to have unique itemsets per transaction.
                This also allows preserving itemset order.
            """
            yield HUIRecord(itemset, itemset_utility)
